### ISSUE TITLE
Deprecate t.conch.insults.colors

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4096]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#4096|
|Type|task|
|Created|2009-11-13 19:41:10Z|

The docstring of [twisted/conch/insults/colors.py](twisted/conch/insults/colors.py) says:

```
You don't really want to use this module. Try helper.py instead.
```

so let's remove this module or deprecate it.

Attachments:

* [4096.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/612/6124cb94c33db249c29395625c4b43deb8be0cb6/f12036c05f15f623a245f0ad820aa5bdb086f9f1.diff) (2424 bytes) - added by cyli on 2010-05-31 04:56:49Z - Deprecate twisted.conch.insults.colors
* [4096.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/612/6124cb94c33db249c29395625c4b43deb8be0cb6/7f1fbc301647b228b0f42169c4f05f5b95e3881c.diff) (2530 bytes) - added by cyli on 2010-06-06 00:04:20Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4096 4096
type__task task
reporter__thijs thijs
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__easy easy
time__1258141270000000 1258141270000000
changetime__1303337233000000 1303337233000000
version__None None
owner__cyli cyli
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @cyli|
|-|-|

  1. Because of the way imports from packages work, the deprecation is first emitted by an `insults.colors`(-like) access which raises an `AttributeError`.  The access is then retried after `colors.py` has actually been loaded; the warning is emitted again, and this time the lookup succeeds and the import is complete.  Normally this will probably be invisible, because duplicate warnings from the same location are suppressed by the reporting code, but trial goes out of its way to make sure every warning gets emitted.
    1. It's probably worth putting something like this into the comment in the test.
    1. It's probably worth filing a ticket for either working around this duplication in `deprecatedModuleAttribute`, or introducing a `deprecatedPackageAttribute` to specifically handle this case.
    1. Of the various ways I've seen to deprecate modules (put a `warnings.warn()` call into the top-level code of the module; deprecate all of the attributes of the module), I think this one may be the nicest.  It may also be worth updating some documentation somewhere to point this out, since it never occurred to me before (and since no other module in Twisted has ever been deprecated in this way, I suppose it didn't occur to anyone else).
  1. I think the preferred form for .removal fragments is _x is now deprecated.  etc_.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli removed owner|
|-|-|

1. Put that entire first paragraph into the comment of the test.
2. Created ticket [#4492](https://github.com/twisted/twisted/issues/4492)
3. Added change to [http://twistedmatrix.com/trac/wiki/CompatibilityPolicy]

And finally, changed the text of the .removal fragment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @cyli|
|-|-|

Thanks.

  1. The `Deprecations` class docstring says something about `client`, but there's no tests or change for `client` that I can see.
  1. Can you add a reference to [#4492](https://github.com/twisted/twisted/issues/4492) in the comment in `test_colors`?
  1. Needs a .removal news fragment

Otherwise looks good.  Merge when the above are fixed.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

Oops, forgot to svn add the news fragment before running diff.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli set status to `closed`|
|-|-|

(In [29213]) Applied patch 4096.2.diff for ticket 4096: Deprecated twisted.conch.insults.colors

Author: cyli
Reviewer: exarkun
Fixes: [#4096](https://github.com/twisted/twisted/issues/4096)

twisted.conch.insults.colors has been deprecated
